### DISCUSSION TITLE
Ordering annot.vertices naturally

### DISCUSSION THREAD

**Discussion by Don-Yin** [pymupdf/PyMuPDF#1742] [Q&A] [answered]
Hi all,

Is there a native way to modify ```annot.vertices``` for highlights in documents with two columns?  I found that ```annot.vertices``` seem to return a top-down ordered list of coordinates by default, which is not great for cross-column highlights,  e.g.:

Is there a native way to modify ```annot.vertices``` for highlights in two-column documents? ```annot.vertices``` appears to return a top-down ordered list of coordinates by default, which is inconvenient for cross-column highlights, this one for example:

<img width="379" alt="example" src="https://user-images.githubusercontent.com/65135356/171621932-d900e6e5-c301-4f69-b9af-7adf6d339df1.png">

In the image above, *area 1* and *2* belong to the same highlight. 

While I am trying to get the rects for each individual 'highlight bars', by doing this:

```python
points = annot.vertices[i * 4 : i * 4 + 4]
quad_count = int(len(annot.vertices) / 4)
for i in range(quad_count):
        points = annot.vertices[i * 4 : i * 4 + 4] 
rect = fitz.Quad(points).rect
```
These bars look like this (using ```get_pixmap```):
It first returns *area 2*:
![0](https://user-images.githubusercontent.com/65135356/171624842-53690ee4-0f96-4472-b697-122710c45065.png)
![1](https://user-images.githubusercontent.com/65135356/171624845-6f71094e-c4bb-495d-b061-e94787ebc7cb.png)
![2](https://user-images.githubusercontent.com/65135356/171624847-aa5c8dd2-037f-4aea-9227-f2169715b0a5.png)
![3](https://user-images.githubusercontent.com/65135356/171624849-5198a79e-27e7-46e2-a83a-b0b8fd914b6f.png)

And *area 1*:
![4](https://user-images.githubusercontent.com/65135356/171624853-e840e405-b389-4141-94d9-e8e1b7ea15d9.png)
![5](https://user-images.githubusercontent.com/65135356/171624856-a1eb67bc-d13a-43d8-9304-e803bc15489e.png)
![6](https://user-images.githubusercontent.com/65135356/171624858-689ae358-c389-4129-8159-7e547c796b15.png)
![7](https://user-images.githubusercontent.com/65135356/171624861-1d333496-03c0-4e23-be6c-e2f62ef1c162.png)
![8](https://user-images.githubusercontent.com/65135356/171624862-e98dd01b-a8ce-4033-92e5-e9f980ab743c.png)

This order is the opposite of what I had hoped for. Again, the issue is that the returning order of vertices appears to have ignored the columns and worked from top to bottom - which is understandable. I understand that this could simply be how highlight is coded.

But still, let me ask: is there a ```fitz``` native way to take PDF columns into account for the vertices? I'm considering writing a clustering algorithm, but that seems like a lot of work which I'd like to avoid if there is a easier way.

I'd appreciate any advice.
Don

**Comment by JorjMcKie** [comment, marked as answer]
No there is not. Consult the PDF spec to see how this information is stored in a PDF file.

**Reply by JorjMcKie** [reply]
Also note that you cannot rely on any order for annotations when retrieving them from `page.annots()`.
Or rather: they are retrieved in the same order as they have been stored by the creator, which is the order as the annotation xrefs are stored in the array `/Annots` present in the page definition. You can access this array via `doc.xref_get_key(page.xref, "Annots")`, or, more reliably(*), via `page.get_annot_xrefs()`.

(*) the second method automatically resolves any indirect references, whereas the first one may return another separate xref, which then contains the array.
